contracts/AnybodyProblem.sol

```diff
@@ -0,0 +1,635 @@
+        input[0],
+        input[1],
+        input[2],
+        input[3],
+        input[4]
+    ];
+    runs[runId].levels[levelIndex].tmpInflightMissile = newOutflightMissile;
+
+    uint256 time = input[5 + bodyCount * 5];
+
+    verifyProof(bodyCount, verifier, a, b, c, input);
+
+    Level memory levelData = runs[runId].levels[levelIndex];
+
+    levelData.time += time;
+    require(levelData.time <= maxTick, "Time limit exceeded");
+
+    uint256 bodiesGone;
+    Body memory bodyData;
+    for (uint256 i = 0; i < bodyCount; i++) {
+        bodyData = levelData.tmpBodyData[i];
+
+        verifyBodyDataMatches(bodyData, input, bodyCount, i);
+        bodyData = extractBodyData(bodyData, input, i);
+
+        if (i == 0) {
+          require(bodyData.radius != 0, "You shot the body you should protect");
+        }
+        if (bodyData.radius == 0) {
+            bodiesGone++;
+        }
+        levelData.tmpBodyData[i] = bodyData;
+    }
+    runs[runId].levels[levelIndex] = levelData;
+
+    if (bodiesGone == level) {
+      runs[runId].levels[levelIndex].solved = true;
+      emit LevelSolved(msg.sender, runId, level, levelData.time, day);
+      runs[runId].accumulativeTime += levelData.time;
+      if (level == LEVELS) {
+        runs[runId].solved = true;
+        require(msg.value == price, "Incorrect payment");
+        (bool sent, bytes memory data) = proceedRecipient.call{
+            value: msg.value
+        }("");
+        Speedruns(speedruns).__mint(runId, msg.sender);
+        emit EthMoved(proceedRecipient, sent, data, msg.value);
+        emit RunSolved(msg.sender, runId, runs[runId].accumulativeTime, day);
+        gamesPlayed[msg.sender].total++;
+        addToLeaderboard(runId);
+
+      } else {
+        addNewLevelData(runId);
+      }
+    }
+  }
+  function addToLeaderboard(uint256 runId) internal {
+    addToFastestByDay(runId);
+    addToLongestStreak(runId);
+    addToMostPlayed(runId);
+  }
+  function addToLongestStreak(uint256 runId) internal {
+    uint256 day = runs[runId].day;
+    Record memory record = gamesPlayed[msg.sender];
+    if (record.lastPlayed + SECONDS_IN_A_DAY != day) {
+      record.streak = 1;
+    } else {
+      record.streak++;
+    }
+    record.lastPlayed = day;
+    gamesPlayed[msg.sender] = record;
+
+    for (uint256 i = 0; i < longestStreak.length; i++) {
+      if (record.streak > gamesPlayed[longestStreak[i]].streak) {
+        for (uint256 j = longestStreak.length - 1; j > i; j--) {
+          longestStreak[j] = longestStreak[j - 1];
+        }
+        longestStreak[i] = msg.sender;
+        break;
+      }
+    }
+  }
+  function addToMostPlayed(uint256 runId) internal {
+    Record memory record = gamesPlayed[msg.sender];
+    for (uint256 i = 0; i < mostGames.length; i++) {
+      if (record.total > gamesPlayed[mostGames[i]].total) {
+        for (uint256 j = mostGames.length - 1; j > i; j--) {
+          mostGames[j] = mostGames[j - 1];
```

Comment:
it looks like you store everything from all-time in the leaderboards.

could we cap gas cost of the insert by limiting to storing top n (say 200)?

Reply:
do you have an idea of how much the gas cost will scale w leaderboard length?

Reply:
there's no real change in cost with larger indexed arrays. Costs the same to store the 1st value as 1,000,000,000th value.

Reply:
ah, i thought it would slowly tick up because as you add more scores you have to spend more time in this loop. I think it's O(n) though